satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,427 @@
+								AddressedOrderLimit: limit,
+								PrivateKey:          privateKey,
+							},
+						},
+					}
+					err = stream.Send(transferMsg)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					pending.put(limit.Limit.PieceId, &pendingTransfer{
+						path:             inc.Path,
+						pieceSize:        pieceSize,
+						satelliteMessage: transferMsg,
+					})
+				}
+			}
+		}
+		return nil
+	})
+
+	for {
+		// if there are no more transfers and the pending queue is empty, send complete
+		if atomic.LoadInt32(&morePiecesFlag) == 0 && pending.length() == 0 {
+			// TODO check whether failure threshold is met before sending completed
+			// TODO needs exit signature
+			transferMsg := &pb.SatelliteMessage{
+				Message: &pb.SatelliteMessage_ExitCompleted{
+					ExitCompleted: &pb.ExitCompleted{},
+				},
+			}
+			err = stream.Send(transferMsg)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			break
+		}
+		// skip if there are none pending
+		if pending.length() == 0 {
+			continue
+		}
+
+		request, err := stream.Recv()
+		if err != nil {
+			return eofHandler(err)
+		}
+
+		switch m := request.GetMessage().(type) {
+		case *pb.StorageNodeMessage_Succeeded:
+			err = endpoint.handleSucceeded(ctx, pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+
+		case *pb.StorageNodeMessage_Failed:
+			err = endpoint.handleFailed(ctx, pending, nodeID, m)
+			if err != nil {
+				return Error.Wrap(err)
+			}
+		default:
+			return Error.New("unknown storage node message: %v", m)
+		}
+	}
+
+	if err := group.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (endpoint *Endpoint) sendPiecesToTransfer(ctx context.Context, stream processStream, nodeID storj.NodeID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	endpoint.log.Debug("sending pieces to transfer.", zap.String("nodeID", nodeID.String()))
+
+	return nil
+}
+
+func (endpoint *Endpoint) handleSucceeded(ctx context.Context, pending *pendingMap, nodeID storj.NodeID, message *pb.StorageNodeMessage_Succeeded) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	if message.Succeeded.GetAddressedOrderLimit() == nil {
+		return Error.New("Addressed order limit cannot be nil.")
+	}
+	if message.Succeeded.GetOriginalPieceHash() == nil {
+		return Error.New("Original piece hash cannot be nil.")
+	}
+	endpoint.log.Debug("transfer succeeded.", zap.String("piece ID", message.Succeeded.AddressedOrderLimit.Limit.PieceId.String()))
+
+	// TODO validation
+
+	transfer, ok := pending.get(message.Succeeded.OriginalPieceHash.PieceId)
+	if !ok {
+		endpoint.log.Debug("could not find transfer message in pending queue. skipping .", zap.String("piece ID", message.Succeeded.AddressedOrderLimit.Limit.PieceId.String()))
+	}
+
+	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, nodeID, transfer.path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	var failed int64
+	if transferQueueItem.FailedCount != nil && *transferQueueItem.FailedCount > 0 {
+		failed = -1
```

Comment:
why are we setting failed to -1 here?

Reply:
It is a little difficult to follow, but my understanding is that we will increment the node failures once for every unique transfer queue item that fails to transfer. Let's say Node A fails to transfer Piece 1 two times, and then successfully transfers it the third time. We will have incremented Node A's graceful exit fail count by 1, even though it ended up successfully transferring the piece in the end. This logic counteracts that.

Reply:
this seems a little convoluted to me, but I'm also not sure of a better way. okay to leave it for now.